### PR TITLE
fix: prevent dupe calls

### PR DESCRIPTION
**Pull request by tlambert03**
trying to fix the duplicate call thing mentioned in https://github.com/pyapp-kit/magicgui/pull/544#pullrequestreview-1335180278

### CODE REVIEW

**Comment by tlambert03** [comment]
ugh... this is just a real annoying problem. 

1. in order to truly support the proper `Optional[type]` return annotation that @Czaki was going for in https://github.com/napari/napari/pull/5595, magicgui needs to move from a simple  `type -> list[callback]` map, to a much more complicated `set[type] -> list[callback]` map... and then when looking up a callback for a return type annotation, look for the first `set[type]` that is a superset of the return callback set.
2. that more complicated map makes everything more complicated:
    - it slows down callback lookup at runtime
    - it makes it much harder to undo type registration (e.g. in the `type_registered` context manager)

basically, this all just further emphasizes that the `return_callback` feature of register_type is far more complicated to do "correctly", and makes me want to deprecate the whole feature altogether as mentioned in #543 ...

I really don't know the best solution here.

**Comment by tlambert03** [comment]
one possibility here is just to temporarily go back to the previous behavior (that I absolutely thing was *incorrect* behavior) of calling callbacks registered for `Type` even when the return annotation of the `@magicgui` function is `Optional[Type]` (as mentioned in https://github.com/pyapp-kit/magicgui/issues/541).

That's a big bummer, cause it's just obviously wrong behavior.  @Czaki, what do you think here?

**Comment by Czaki** [comment]
Why not just check presence of callback in lust before append? Registration type should be rare operation and such list should be short.

**Comment by tlambert03** [comment]
Because if we're dealing with sets as keys you now have to check all keys for a subset match, remember which ones matched, etc...

**Comment by tlambert03** [comment]
It's not that it's not possible, it's just getting uglier and uglier for something that was an accidental "feature/bug" to begin with

**Comment by Czaki** [comment]
Because of `type2callback` return first callback list where superset is found this test will fail:

```
def test_proper_call():
    @magicgui
    def func(a: int, b: int) -> int:
        return a + b

    mock1 = Mock()
    mock2 = Mock()

    register_type(int, return_callback=mock1)
    register_type(Optional[int], return_callback=mock2)

    func(1, 2)

    mock2.assert_called_once()
    mock1.assert_called_once()
```

**Comment by tlambert03** [comment]
closing, see https://github.com/pyapp-kit/magicgui/pull/546
